clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3575,17 +3575,24 @@ class SyclKernelIntHeaderCreator : public SyclKernelFieldHandler {
 
       Header.addParamDesc(SYCLIntegrationHeader::kind_accessor, Info,
                           CurOffset + offsetOf(FD, FieldTy));
+    } else if (isSyclType(FieldTy, SYCLTypeAttr::stream)) {
+      addParam(FD, FieldTy, SYCLIntegrationHeader::kind_stream);
     } else {
-      if (getMethodByName(ClassTy, FinalizeMethodName))
-        addParam(FD, FieldTy, SYCLIntegrationHeader::kind_stream);
-      else {
-        CXXMethodDecl *InitMethod = getMethodByName(ClassTy, InitMethodName);
-        assert(InitMethod && "type must have __init method");
-        const ParmVarDecl *SamplerArg = InitMethod->getParamDecl(0);
-        assert(SamplerArg && "Init method must have arguments");
-        addParam(SamplerArg->getType(), SYCLIntegrationHeader::kind_sampler,
-                 offsetOf(FD, FieldTy));
+      CXXMethodDecl *InitMethod = getMethodByName(ClassTy, InitMethodName);
+      assert(InitMethod && "type must have __init method");
+      const ParmVarDecl *InitArg = InitMethod->getParamDecl(0);
+      assert(InitArg && "Init method must have arguments");
+      QualType T = InitArg->getType();
+      if (isSyclType(FieldTy, SYCLTypeAttr::sampler)) {
+        addParam(T, SYCLIntegrationHeader::kind_sampler, offsetOf(FD, FieldTy));
+        return true;
       }
+      // Add a default case for classes marked with sycl_special_class
+      // attribute.
+      addParam(T,
+               T->isPointerType() ? SYCLIntegrationHeader::kind_pointer
```

Comment:
This thing concerns me a bit. `kind_pointer` and `kind_standard_layout` will likely mean for RT that it should pass a single value which is pointer or some standard layout type to the kernel. What if some new generic special type has several `__init` method parameters? I believe it will be a runtime error.

Reply:
Yea I thought about that. I assume we will require a new 'kind' to be defined in those cases so that the runtime knows it needs to process multiple openCL kernel arguments, the way it does accessors and streams. I'll add an unreachable here like I suggested in an earlier comment since both you and @premanandrao, and even I have some concerns about this.  I guess its safer than specifying default behavior when we don't really know what 'default' can be.

Reply:
Done